### PR TITLE
Deprecate `ServiceDiscoveryRetryStrategy`

### PR DESCRIPTION
**Pull request by chemicL**
Motivation:

`ServiceDiscoveryRetryStrategy` has two responsibilities. One,
implied by the name is retrying errors from the `Publisher`
probided by `ServiceDiscoverer#discover` method. The other, implicit, is
deduplication logic for incoming events. As `RoundRobinLoadBalancer`
recently gained the ability to deduplicate events on its own, this
second responsibility is no longer a requirement. At the same time,
retries can be achieved by `Completable#retryWhen` mechanism using
standard utilities provided by `RetryStrategies`. Therefore, the
intention is to eventually remove `ServiceDiscoveryRetryStrategy` and
its' implementation, `DefaultServiceDiscoveryRetryStrategy`.

Modifications:

- Deprecated `ServiceDiscoveryRetryStrategy` and
  `DefaultServiceDiscoveryRetryStrategy`,
- Deprecated
  `PartitionedHttpClientBuilder#retryServiceDiscoveryErrors(ServiceDiscoveryRetryStrategy)`
  and introduced a variant accepting `BiIntFunction<Throwable, ? extends
  Completable>`,
- Deprecated
  `SingleAddressHttpClientBuilder#retryServiceDiscoveryErrors(ServiceDiscoveryRetryStrategy)`
  and introduced a variant accepting `BiIntFunction<Throwable, ? extends
  Completable>`.

Result:

Complicated and unnecessary `ServiceDiscoveryRetryStrategy` can be
removed in the future and users can rely on standard mechanisms for
retrying `ServiceDiscoverer` errors.

### CODE REVIEW

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java [hunk 1]
```diff
@@ -89,7 +95,13 @@ private DefaultServiceDiscoveryRetryStrategy(final UnaryOperator<E> flipAvailabi
      *
      * @param <ResolvedAddress> The type of address after resolution.
      * @param <E> Type of {@link ServiceDiscovererEvent}s published from {@link ServiceDiscoverer#discover(Object)}.
+     * @see io.servicetalk.concurrent.api.RetryStrategies
+     * @see SingleAddressHttpClientBuilder#retryServiceDiscoveryErrors(BiIntFunction)
+     * @deprecated Use standard mechanisms for retrying using
+     * {@link io.servicetalk.concurrent.api.Completable#retryWhen(BiIntFunction)} on the {@link Publisher} returned from
+     * {@link ServiceDiscoverer#discover(Object)} instead.
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java:102]: This should suggest using client builder methods

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/PartitionedHttpClientBuilder.java [hunk 2]
```diff
@@ -93,13 +95,37 @@
...
+    }
+
+    /**
+     * Sets a retry strategy to retry errors emitted by {@link ServiceDiscoverer}.
+     * <p>
+     * Note, calling this method will unset the value provided via
+     * {@link #retryServiceDiscoveryErrors(ServiceDiscoveryRetryStrategy)} if it was called before.
      * @param retryStrategy a retry strategy to retry errors emitted by {@link ServiceDiscoverer}.
      * @return {@code this}.
      * @see DefaultServiceDiscoveryRetryStrategy.Builder
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/PartitionedHttpClientBuilder.java:121]: Remove reference to `DefaultServiceDiscoveryRetryStrategy`

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java [hunk 3]
```diff
@@ -288,13 +290,37 @@
...
+    }
+
+    /**
+     * Sets a retry strategy to retry errors emitted by {@link ServiceDiscoverer}.
      * @param retryStrategy a retry strategy to retry errors emitted by {@link ServiceDiscoverer}.
+     * <p>
+     * Note, calling this method will unset the value provided via
+     * {@link #retryServiceDiscoveryErrors(ServiceDiscoveryRetryStrategy)} if it was called before.
      * @return {@code this}.
      * @see DefaultServiceDiscoveryRetryStrategy.Builder
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:316]: Remove reference to `DefaultServiceDiscoveryRetryStrategy`

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java [hunk 4]
```diff
@@ -225,12 +235,18 @@ StreamingHttpClient build() {
...
             return new RetryingServiceDiscoverer<>(new StatusAwareServiceDiscoverer<>(sd, sdStatus),
-                            serviceDiscovererRetryStrategy == null ?
-                                    DefaultServiceDiscoveryRetryStrategy.Builder
-                                            .<R>withDefaults(executionContext.executor(),
-                                                    SD_RETRY_STRATEGY_INIT_DURATION, SD_RETRY_STRATEGY_JITTER).build() :
-                                    serviceDiscovererRetryStrategy);
+                    sdRetryStrategy,
+                    deprecatedServiceDiscovererRetryStrategy == null ?
+                            DefaultServiceDiscoveryRetryStrategy.Builder
+                                    .<R>withDefaults(executionContext.executor(),
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:247]: We should avoid building `DefaultServiceDiscoveryRetryStrategy`, it should be used only if a custom one is provided by users.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java [hunk 5]
```diff
@@ -723,17 +749,24 @@ void resetError() {
...
+        private final ServiceDiscoveryRetryStrategy<R, E> deprecatedRetryStrategy;
 
         RetryingServiceDiscoverer(final ServiceDiscoverer<U, R, E> delegate,
-                                  final ServiceDiscoveryRetryStrategy<R, E> retryStrategy) {
+                                  @Nullable final BiIntFunction<Throwable, ? extends Completable> retryStrategy,
+                                  final ServiceDiscoveryRetryStrategy<R, E> deprecatedRetryStrategy) {
             super(delegate);
-            this.retryStrategy = requireNonNull(retryStrategy);
+            this.retryStrategy = retryStrategy;
+            this.deprecatedRetryStrategy = requireNonNull(deprecatedRetryStrategy);
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:761]: Consider asserting that only one of the retry strategies is not null

**Review by idelpivnitskiy — APPROVED** [review]
LGTM after comments addressed

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java (hunk 5 above)
Comment (chemicL) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:761]: I added validation in the constructor.

**Comment by chemicL** [comment]
0.41: ebe29110c62a69cb48fbc4c636c9d4036efcf711

**Pull request merged**